File: vault/audit.py
"""
Structured audit logging for Rune-Vault operations.

Emits one JSON line per gRPC request to a dedicated audit log,
separate from the application log. Supports file-based daily rotation
and stdout JSON mode for container environments.

Configuration via VAULT_AUDIT_LOG env var:
  (empty)        disabled
  file           /var/log/rune-vault/audit.log, daily rotation, 30-day retention
  file:/path     custom file path
  stdout         JSON lines to stdout
  file+stdout    both
"""

import json
import logging
import os
import sys
from logging.handlers import TimedRotatingFileHandler
from typing import Any

_DEFAULT_AUDIT_PATH = "/var/log/rune-vault/audit.log"

# ---------------------------------------------------------------------------
# Configuration parsing
# ---------------------------------------------------------------------------


def _parse_audit_config(env_value: str) -> dict:
    """Parse VAULT_AUDIT_LOG into {"file": path | None, "stdout": bool}."""
    if not env_value:
        return {"file": None, "stdout": False}

    parts = [p.strip() for p in env_value.split("+")]
    config: dict[str, Any] = {"file": None, "stdout": False}

    for part in parts:
        lowered = part.lower()
        if lowered == "stdout":
            config["stdout"] = True
        elif lowered == "file":
            config["file"] = _DEFAULT_AUDIT_PATH
        elif lowered.startswith("file:"):
            config["file"] = part.split(":", 1)[1].strip()

    return config


# ---------------------------------------------------------------------------
# Source IP extraction
# ---------------------------------------------------------------------------


def extract_source_ip(context) -> str:
    """Extract client IP from gRPC context.peer().

    peer() returns strings like:
      'ipv4:10.0.0.1:12345'
      'ipv6:[::1]:12345'
      'unix:/path/to/socket'
    """
    try:
        peer = context.peer()
        if peer is None:
            return "unknown"
        if peer.startswith("ipv4:"):
            # ipv4:10.0.0.1:12345 -> 10.0.0.1
            return peer[5:].rsplit(":", 1)[0]
        if peer.startswith("ipv6:"):
            addr = peer[5:]
            if addr.startswith("["):
                # [::1]:12345 -> [::1]
                return addr.split("]", 1)[0] + "]"
            return addr.rsplit(":", 1)[0]
        return peer
    except Exception:
        return "unknown"


# ---------------------------------------------------------------------------
# AuditLogger
# ---------------------------------------------------------------------------


class AuditLogger:
    """JSON-structured audit logger with file rotation and stdout support."""

    def __init__(self, config: dict):
        self._logger = logging.getLogger("rune.vault.audit")
        self._logger.setLevel(logging.INFO)
        self._logger.propagate = False

        # Close and remove any pre-existing handlers (e.g. during tests)
        for h in self._logger.handlers[:]:
            h.close()
            self._logger.removeHandler(h)

        if config.get("file"):
            handler = TimedRotatingFileHandler(
                config["file"],
                when="midnight",
                backupCount=30,
                utc=True,
            )
            handler.setFormatter(logging.Formatter("%(message)s"))
            self._logger.addHandler(handler)

        if config.get("stdout"):
            handler = logging.StreamHandler(sys.stdout)
            handler.setFormatter(logging.Formatter("%(message)s"))
            self._logger.addHandler(handler)

    @property
    def enabled(self) -> bool:
        return len(self._logger.handlers) > 0

    def log(
        self,
        *,
        timestamp: str,
        user_id: str,
        method: str,
        top_k: int | None,
        result_count: int,
        status: str,
        source_ip: str,
        latency_ms: float,
        error: str | None = None,
    ) -> dict:
        """Emit a single structured audit entry. Returns the entry dict."""
        entry: dict[str, Any] = {
            "timestamp": timestamp,
            "user_id": user_id,
            "method": method,
            "top_k": top_k,
            "result_count": result_count,
            "status": status,
            "source_ip": source_ip,
            "latency_ms": round(latency_ms, 2),
        }
        if error is not None:
            entry["error"] = error
        self._logger.info(json.dumps(entry, separators=(",", ":")))
        return entry


# ---------------------------------------------------------------------------
# Module-level singleton
# ---------------------------------------------------------------------------

_config = _parse_audit_config(os.environ.get("VAULT_AUDIT_LOG", ""))
audit_logger = AuditLogger(_config)

File: vault/internal/server/audit.go
package server

import (
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"google.golang.org/grpc/peer"
	"gopkg.in/natefinch/lumberjack.v2"
)

// AuditMode parses AuditConfig.Mode strings into per-sink booleans.
// Valid values: "", "file", "stdout", "file+stdout".
type AuditMode struct {
	File   bool
	Stdout bool
}

func ParseAuditMode(mode string) AuditMode {
	out := AuditMode{}
	if mode == "" {
		return out
	}
	for _, p := range strings.Split(mode, "+") {
		switch strings.TrimSpace(strings.ToLower(p)) {
		case "file":
			out.File = true
		case "stdout":
			out.Stdout = true
		}
	}
	return out
}

// AuditEntry is the JSON structure written per request. Fields and order
// must match vault/audit.py:118-145 to keep golden compat tests aligned.
type AuditEntry struct {
	Timestamp   string  `json:"timestamp"`
	UserID      string  `json:"user_id"`
	Method      string  `json:"method"`
	TopK        *int32  `json:"top_k"`
	ResultCount int     `json:"result_count"`
	Status      string  `json:"status"`
	SourceIP    string  `json:"source_ip"`
	LatencyMs   float64 `json:"latency_ms"`
	Error       *string `json:"error,omitempty"`
}

// AuditLogger writes structured audit entries. Closed loggers are no-ops.
type AuditLogger struct {
	mu      sync.Mutex
	writers []io.Writer
	closers []io.Closer
}

// NewAuditLogger constructs a logger for the given mode + file path.
// Returns a logger with Enabled() == false when the mode is empty.
func NewAuditLogger(cfg AuditConfig) (*AuditLogger, error) {
	mode := ParseAuditMode(cfg.Mode)
	l := &AuditLogger{}

	if mode.File {
		path := cfg.Path
		if path == "" {
			path = "/var/log/rune-vault/audit.log"
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o700); err != nil {
			return nil, fmt.Errorf("audit: mkdir log dir: %w", err)
		}
		// Lumberjack handles daily-ish rotation by size + age. Match the
		// Python deployment's 30-day retention; size cap is high enough
		// that audit volume drives rotation by age, not size.
		rot := &lumberjack.Logger{
			Filename:   path,
			MaxSize:    100, // MB
			MaxAge:     30,
			MaxBackups: 30,
			LocalTime:  false,
			Compress:   false,
		}
		l.writers = append(l.writers, rot)
		l.closers = append(l.closers, rot)
	}

	if mode.Stdout {
		l.writers = append(l.writers, os.Stdout)
	}
	return l, nil
}

// Enabled reports whether at least one sink is configured.
func (a *AuditLogger) Enabled() bool {
	if a == nil {
		return false
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	return len(a.writers) > 0
}

// Log emits a single audit entry. Round-trip latency is rounded to 2dp
// to match Python's `round(latency_ms, 2)`.
func (a *AuditLogger) Log(e AuditEntry) {
	if a == nil || !a.Enabled() {
		return
	}
	e.LatencyMs = roundTo(e.LatencyMs, 2)

	buf, err := json.Marshal(&e)
	if err != nil {
		return
	}
	buf = append(buf, '\n')

	a.mu.Lock()
	defer a.mu.Unlock()
	for _, w := range a.writers {
		_, _ = w.Write(buf)
	}
}

// Close flushes file writers and prevents future Log calls from writing.
func (a *AuditLogger) Close() error {
	if a == nil {
		return nil
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	var firstErr error
	for _, c := range a.closers {
		if err := c.Close(); err != nil && firstErr == nil {
			firstErr = err
		}
	}
	a.writers = nil
	a.closers = nil
	return firstErr
}

func roundTo(v float64, decimals int) float64 {
	mult := 1.0
	for i := 0; i < decimals; i++ {
		mult *= 10
	}
	if v >= 0 {
		return float64(int64(v*mult+0.5)) / mult
	}
	return float64(int64(v*mult-0.5)) / mult
}

// ExtractSourceIP mirrors vault/audit.py:55-78 — peer addresses come in
// gRPC's "ipv4:H:P", "ipv6:[::1]:P", or "unix:/path" form.
func ExtractSourceIP(p *peer.Peer) string {
	if p == nil || p.Addr == nil {
		return "unknown"
	}
	addr := p.Addr.String()
	switch a := p.Addr.(type) {
	case *net.TCPAddr:
		if a.IP == nil {
			return addr
		}
		return a.IP.String()
	case *net.UnixAddr:
		return "unix:" + a.Name
	}
	if h, _, err := net.SplitHostPort(addr); err == nil {
		return h
	}
	return addr
}

func nowUTCISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000000Z07:00")
}
